### PR TITLE
Adding support for mordents and turns

### PR DESCRIPTION
**Pull request by rettinghaus**
This PR adds support for mordents and turns for MEI and MusicXML

ornaments

MusicXML:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE score-partwise PUBLIC "-//Recordare//DTD MusicXML 3.0 Partwise//EN" "http://www.musicxml.org/dtds/partwise.dtd">
<score-partwise>
  <identification>
    <encoding>
      <software>MuseScore 2.0.3.1</software>
      <encoding-date>2017-02-01</encoding-date>
      <supports element="accidental" type="yes"/>
      <supports element="beam" type="yes"/>
      <supports element="print" attribute="new-page" type="yes" value="yes"/>
      <supports element="print" attribute="new-system" type="yes" value="yes"/>
      <supports element="stem" type="yes"/>
      </encoding>
    </identification>
  <defaults>
    <scaling>
      <millimeters>7.05556</millimeters>
      <tenths>40</tenths>
      </scaling>
    <page-layout>
      <page-height>1683.36</page-height>
      <page-width>1190.88</page-width>
      <page-margins type="even">
        <left-margin>56.6929</left-margin>
        <right-margin>56.6929</right-margin>
        <top-margin>56.6929</top-margin>
        <bottom-margin>113.386</bottom-margin>
        </page-margins>
      <page-margins type="odd">
        <left-margin>56.6929</left-margin>
        <right-margin>56.6929</right-margin>
        <top-margin>56.6929</top-margin>
        <bottom-margin>113.386</bottom-margin>
        </page-margins>
      </page-layout>
    <word-font font-family="FreeSerif" font-size="10"/>
    <lyric-font font-family="FreeSerif" font-size="11"/>
    </defaults>
  <part-list>
    <score-part id="P1">
      <score-instrument id="P1-I1">
        <instrument-name>Piano</instrument-name>
        </score-instrument>
      <midi-device id="P1-I1" port="1"></midi-device>
      <midi-instrument id="P1-I1">
        <midi-channel>1</midi-channel>
        <midi-program>1</midi-program>
        <volume>78.7402</volume>
        <pan>0</pan>
        </midi-instrument>
      </score-part>
    </part-list>
  <part id="P1">
    <measure number="1" width="566.47">
      <print>
        <system-layout>
          <system-margins>
            <left-margin>0.00</left-margin>
            <right-margin>-0.00</right-margin>
            </system-margins>
          <top-system-distance>70.00</top-system-distance>
          </system-layout>
        </print>
      <attributes>
        <divisions>1</divisions>
        <key>
          <fifths>-1</fifths>
          </key>
        <time>
          <beats>3</beats>
          <beat-type>4</beat-type>
          </time>
        <clef>
          <sign>F</sign>
          <line>4</line>
          </clef>
        </attributes>
      <note default-x="84.86" default-y="-20.00">
        <pitch>
          <step>D</step>
          <octave>3</octave>
          </pitch>
        <duration>1</duration>
        <voice>1</voice>
        <type>quarter</type>
        <stem>down</stem>
        <notations>
          <ornaments>
            <inverted-turn/>
            </ornaments>
          </notations>
        </note>
      <note default-x="244.86" default-y="-25.00">
        <pitch>
          <step>C</step>
          <octave>3</octave>
          </pitch>
        <duration>1</duration>
        <voice>1</voice>
        <type>quarter</type>
        <stem>up</stem>
        <notations>
          <ornaments>
            <turn/>
            </ornaments>
          </notations>
        </note>
      <note default-x="404.87" default-y="-15.00">
        <pitch>
          <step>E</step>
          <octave>3</octave>
          </pitch>
        <duration>1</duration>
        <voice>1</voice>
        <type>quarter</type>
        <stem>down</stem>
        <notations>
          <ornaments>
            <inverted-mordent/>
            <trill-mark/>
            </ornaments>
          </notations>
        </note>
      </measure>
    <measure number="2" width="511.02">
      <note default-x="12.00" default-y="-30.00">
        <pitch>
          <step>B</step>
          <alter>-1</alter>
          <octave>2</octave>
          </pitch>
        <duration>1</duration>
        <voice>1</voice>
        <type>quarter</type>
        <stem>up</stem>
        <notations>
          <ornaments>
            <mordent/>
            </ornaments>
          </notations>
        </note>
      <note default-x="174.81" default-y="-10.00">
        <pitch>
          <step>F</step>
          <octave>3</octave>
          </pitch>
        <duration>1</duration>
        <voice>1</voice>
        <type>quarter</type>
        <stem>down</stem>
        <notations>
          <ornaments>
            <inverted-mordent long="yes"/>
            </ornaments>
          </notations>
        </note>
      <note default-x="337.61" default-y="-50.00">
        <pitch>
          <step>E</step>
          <octave>2</octave>
          </pitch>
        <duration>1</duration>
        <voice>1</voice>
        <type>quarter</type>
        <stem>up</stem>
        <notations>
          <fermata type="upright"/>
          </notations>
        </note>
      <barline location="right">
        <bar-style>light-heavy</bar-style>
        </barline>
      </measure>
    </part>
  </score-partwise>
```

MEI:
```xml
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="3.0.0">
  <meiHead>
    <fileDesc>
      <titleStmt>
        <title />
      </titleStmt>
      <pubStmt>
      </pubStmt>
    </fileDesc>
    <encodingDesc>
      <appInfo>
        <application isodate="2017-02-01T22:34:41" version="0.9.14-dev-666e967-dirty">
          <name>Verovio</name>
          <p>Transcoded from MusicXML</p>
        </application>
      </appInfo>
    </encodingDesc>
  </meiHead>
  <music>
    <body>
      <mdiv>
        <score>
          <scoreDef xml:id="scoredef-000000186056400">
            <staffGrp xml:id="m-000000190400415">
              <staffDef xml:id="staffdef-000000091734882" clef.shape="F" clef.line="4" key.sig="1f" meter.count="3" meter.unit="4" n="1" lines="5" />
            </staffGrp>
          </scoreDef>
          <section xml:id="section-000000151805853">
            <measure xml:id="measure-000000179712424" n="1">
              <staff xml:id="staff-000000135558654" n="1">
                <layer xml:id="layer-000000068910121" n="1">
                  <note xml:id="note-000000034477862" dur="4" oct="3" pname="d" stem.dir="down" />
                  <note xml:id="note-000000015859858" dur="4" oct="3" pname="c" stem.dir="up" />
                  <note xml:id="note-000000117549988" dur="4" oct="3" pname="e" stem.dir="down" />
                </layer>
              </staff>
              <turn xml:id="turn-000000078338673" staff="1" startid="#note-000000034477862" form="inv" />
              <turn xml:id="turn-000000053914494" staff="1" startid="#note-000000015859858" form="norm" />
              <mordent xml:id="mordent-000000192453205" staff="1" startid="#note-000000117549988" form="inv" />
              <trill xml:id="trill-000000021157407" staff="1" startid="#note-000000117549988" />
            </measure>
            <measure xml:id="measure-000000184009316" n="2" right="end">
              <staff xml:id="staff-000000207660792" n="1">
                <layer xml:id="layer-000000064519804" n="1">
                  <note xml:id="note-000000119860861" dur="4" oct="2" pname="b" stem.dir="up">
                    <accid xml:id="accid-000000161833302" accid.ges="f" />
                  </note>
                  <note xml:id="note-000000097925591" dur="4" oct="3" pname="f" stem.dir="down" />
                  <note xml:id="note-000000204800065" dur="4" oct="2" pname="e" stem.dir="up" />
                </layer>
              </staff>
              <mordent xml:id="mordent-000000144274511" staff="1" startid="#note-000000119860861" form="norm" />
              <mordent xml:id="mordent-000000003947599" staff="1" startid="#note-000000097925591" form="inv" long="true" />
              <fermata xml:id="fermata-000000087909766" staff="1" startid="#note-000000204800065" form="norm" place="above" />
            </measure>
          </section>
        </score>
      </mdiv>
    </body>
  </music>
</mei>
```

### CODE REVIEW

**Comment by lpugin** [comment]
Perfect!

File: src/iomusxml.cpp [hunk 1]
```diff
@@ -1323,6 +1325,44 @@ void MusicXmlInput::ReadMusicXmlNote(pugi::xml_node node, Measure *measure, int
...
+    if (xmlMordentInv) {
+        Mordent *mordent = new Mordent();
+        m_controlElements.push_back(std::make_pair(measureNum, mordent));
+        mordent->SetStaff(staff->Att::StrToXsdPositiveIntegerList(std::to_string(staff->GetN())));
+        mordent->SetStartid("#" + element->GetUuid());
+        // color
+        std::string colorStr = GetAttributeValue(xmlMordentInv.node(), "color");
+        if (!colorStr.empty()) mordent->SetColor(colorStr.c_str());
+        // form
+        mordent->SetForm(mordentLog_FORM_inv);
```
Comment (lpugin) [inline src/iomusxml.cpp]: Actually one comment: do you really want this by default? Couldn't it remain unset (NONE)?

File: src/iomusxml.cpp [hunk 2]
```diff
@@ -1338,6 +1378,38 @@ void MusicXmlInput::ReadMusicXmlNote(pugi::xml_node node, Measure *measure, int
...
+    if (xmlTurn) {
+        Turn *turn = new Turn();
+        m_controlElements.push_back(std::make_pair(measureNum, turn));
+        turn->SetStaff(staff->Att::StrToXsdPositiveIntegerList(std::to_string(staff->GetN())));
+        turn->SetStartid("#" + element->GetUuid());
+        // color
+        std::string colorStr = GetAttributeValue(xmlTurn.node(), "color");
+        if (!colorStr.empty()) turn->SetColor(colorStr.c_str());
+        // form
+        turn->SetForm(turnLog_FORM_norm);
```
Comment (lpugin) [inline src/iomusxml.cpp]: Same here. If possible I would leave it unset

File: src/iomusxml.cpp (hunk 1 above)
Comment (rettinghaus) [inline src/iomusxml.cpp]: I thought about this. But MusicXML uses for inverted turns and mordents special elements. So basically converting to MEI means merging two elements into one, that's why I set the form attribute in every case.

Comment (lpugin) [inline src/iomusxml.cpp]: I see. I leave it up to you since you know MusicXML far better than me.

**Pull request merged**